counsel.el

```diff
@@ -4517,6 +4518,53 @@ If there is no such buffer, start a new `shell' with NAME."
                             (reusable-frames . visible)))
     (shell name)))
 
+;;** `counsel-firefox-bookmarks`
+(defvar counsel-firefox-bookmarks-html-file
+  (car (file-expand-wildcards "~/.mozilla/firefox/*/bookmarks.html"))
+  "Firefox's auto exported html bookmarks file.")
+
+(defun counsel-firefox-bookmarks-action (x)
+  "Browse candidate X."
+  (browse-url (get-text-property 0 'href x)))
+
+(defun counsel-firefox-bookmarks--candidates ()
+  (let ((candidates))
+    (if (and counsel-firefox-bookmarks-html-file (file-exists-p counsel-firefox-bookmarks-html-file))
+      (with-temp-buffer
+        (insert-file-contents counsel-firefox-bookmarks-html-file)
+        (goto-char (point-min))
+        (while (re-search-forward "*?<A HREF=\"\\([^\"]+\\)\"[^>]*>\\([^<]+\\)</A>" nil t)
+          (let ((a (match-string 0))
+                (href (match-string 1))
+                (text (match-string 2))
+                (tags nil))
+            (if (string-match "TAGS=\"\\([^\"]+\\)\"" a)
+                (setq tags (match-string 1 a)))
+            (push (propertize (format "%s%s" text (if tags (concat "    :" (replace-regexp-in-string "," ":" tags) ":") "")) 'href href) candidates))))
+      (warn "`counsel-firefox-bookmarks-html-file` not exists"))
+      candidates))
```

Comment:
Given my comments above, I suggest the following refactor:
```el

(defun counsel-firefox-bookmarks--candidates ()
  "Return list of `counsel-firefox-bookmarks' candidates."
  (unless (and counsel-firefox-bookmarks-html-file
               (file-readable-p counsel-firefox-bookmarks-html-file))
    (signal 'file-error (list "Opening `counsel-firefox-bookmarks-html-file'"
                              "No such readable file"
                              counsel-firefox-bookmarks-html-file)))
  (require 'xml)
  (with-temp-buffer
    (let ((case-fold-search t)
          candidates)
      (insert-file-contents counsel-firefox-bookmarks-html-file)
      (while (re-search-forward
              "<a href=\"\\([^\"]+?\\)\"[^>]*?>\\([^<]+?\\)</a>" nil t)
        (let* ((a (match-string 0))
               (href (match-string 1))
               (text (save-match-data
                       (xml-substitute-special (match-string 2))))
               (tags (and (string-match "tags=\"\\([^\"]+?\\)\"" a)
                          (concat " :" (subst-char-in-string
                                        ?, ?: (match-string 1 a) t)
                                  ":")))
               (cand (if tags (concat text tags) text)))
          (put-text-property 0 (length cand) 'href href cand)
          (push cand candidates)))
      candidates)))
```

In the future, IWBN if we used `libxml-parse-html-region` (available since Emacs 24) when available, instead of parsing HTML with regexps.